src/snowflake/snowpark/session.py

```diff
@@ -2276,6 +2303,30 @@ def create_dataframe(
                 )
                 sf_schema = self._conn._get_current_parameter("schema", quoted=False)
 
+                # If the user specifies schema for their dataframe, we try out best to match
+                # it by create a temp table with the specified schema, and load the data into
+                # the temp table. If we fail, go back to old method using infer schema.
+                if isinstance(
+                    schema, StructType
+                ) and self._create_temp_table_for_given_schema(temp_table_name, schema):
```

Comment:
should we consider dropping this temp table after evaluation?

Reply:
If we are successful at writing pandas data into the temp table, then we will use this table in this session. If we drop it, we won't be able to use the dataframe.

Reply:
Now that we have temp table clean-up, we can update `Session.write_pandas` to clean-up temp table once the usage is done?

Reply:
I separated this change out into this PR: https://github.com/snowflakedb/snowpark-python/pull/2784